crates/ruff_linter/src/noqa.rs

```diff
@@ -418,182 +300,359 @@ impl<'a> FileNoqaDirectives<'a> {
     }
 }
 
-/// An individual file-level exemption (e.g., `# ruff: noqa` or `# ruff: noqa: F401, F841`). Like
-/// [`FileNoqaDirectives`], but only for a single line, as opposed to an aggregated set of exemptions
-/// across a source file.
+/// Output of lexing a `noqa` directive.
 #[derive(Debug)]
-pub(crate) enum ParsedFileExemption<'a> {
-    /// The file-level exemption ignores all rules (e.g., `# ruff: noqa`).
-    All,
-    /// The file-level exemption ignores specific rules (e.g., `# ruff: noqa: F401, F841`).
-    Codes(Codes<'a>),
+pub(crate) struct NoqaLexerOutput<'a> {
+    warnings: Vec<LexicalWarning>,
+    directive: Directive<'a>,
 }
 
-impl<'a> ParsedFileExemption<'a> {
-    /// Return a [`ParsedFileExemption`] for a given `comment_range` in `source`.
-    fn try_extract(comment_range: TextRange, source: &'a str) -> Result<Option<Self>, ParseError> {
-        let line = &source[comment_range];
-        let offset = comment_range.start();
-        let init_line_len = line.text_len();
+/// Lexes in-line `noqa` comment, e.g. `# noqa: F401`
+pub(crate) fn lex_inline_noqa(
+    comment_range: TextRange,
+    source: &str,
+) -> Result<Option<NoqaLexerOutput<'_>>, LexicalError> {
+    let line = &source[comment_range];
+    let offset = comment_range.start();
+
+    for (char_index, _) in line
+        .char_indices()
+        .filter(|(pos, c)| matches!(c, 'n' | 'N') && is_noqa_at_position(line, *pos))
+    {
+        let noqa_literal_end = char_index + "noqa".len();
+
+        // Find comment start (position of '#')
+        let before_noqa = &line[..char_index].trim_end();
+        if !before_noqa.ends_with('#') {
+            continue;
+        }
 
-        let line = Self::lex_whitespace(line);
-        let Some(line) = Self::lex_char(line, '#') else {
-            return Ok(None);
-        };
-        let comment_start = init_line_len - line.text_len() - '#'.text_len();
-        let line = Self::lex_whitespace(line);
+        // Calculate positions
+        let comment_start = TextSize::try_from(before_noqa.len() - '#'.len_utf8()).unwrap();
+        let noqa_literal_end = TextSize::try_from(noqa_literal_end).unwrap();
 
-        let Some(line) = Self::lex_flake8(line).or_else(|| Self::lex_ruff(line)) else {
-            return Ok(None);
-        };
+        let lexer = NoqaLexer::new(
+            source,
+            TextRange::new(offset + noqa_literal_end, comment_range.end()),
+        );
+        return Ok(Some(lexer.lex(offset + comment_start)?));
```

Comment:
What's the reason that we return the first code here instead of collecting all? Do we not support `# noqa: RUF100 # noqa: RUF101`